### PR TITLE
[Chrony Receiver] Part 2 - Internal chrony client

### PR DESCRIPTION
**Pull request by MovieStoreGuy**
**Description:** 
Implements the client will be used here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12101

**Link to tracking Issue:**

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/11789

**Testing:** 

Client comes with testing

**Documentation:** 

Is an internal implementation, so no behaviour change to document

### CODE REVIEW

**Comment by jpkrohling** [comment]
Once the part 1 is merged, please rebase this and ping me for review.

**Comment by MovieStoreGuy** [comment]
The one thing that I have notice is that windows support is more challenging than what I expected.

Is there a practice for making an OS experimental for a given receiver or reducing functionality?

**Comment by jpkrohling** [comment]
Take a look at the journald receiver: it has a nolinux file which could serve as model for your case: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/journaldreceiver

**Comment by MovieStoreGuy** [comment]
Eugh... Looks like I have made a fool of the git history. Let me try resolve this locally

**Comment by MovieStoreGuy** [comment]
Ummm, I did not mean to ping everyone to review my PR. I don't need that much attention.

**Comment by MovieStoreGuy** [comment]
@jpkrohling , @dmitryax ,

If you have a chance to review, would be greatly appreciated.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/client.go:65]: What's this used for? Can't you just use the UNIX timestamp instead of a new random number?

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/client.go:86]: Can GetTrackingData be called concurrently? If so, this needs to be an atomic operation.

File: receiver/chronyreceiver/internal/chrony/client_test.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/client_test.go:145]: You can use values waaay lower. Like, a timeout of 10ms and a sleep of 50ms. Waiting 2s for nothing isn't going to do well for our already long CI wait times.

File: receiver/chronyreceiver/internal/chrony/types.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/types.go:31]: Then you need to keep the copyright headers. Just add the original ones to the top of this source file, keeping the OpenTelemetry ones, as you made changes as well.

File: receiver/chronyreceiver/internal/chrony/types.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/types.go:32]: Use a permanent link for this.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (MovieStoreGuy) [reply receiver/chronyreceiver/internal/chrony/client.go:56]: This is my bad, I had forgot to fix this up

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (MovieStoreGuy) [reply receiver/chronyreceiver/internal/chrony/client.go:59]: So the issue with `url.Parse` is it doesn't work for windows like directory name like `unix://c:\\path\chrony.sock`.

Splitting based on `://` at least meant that if any user was running chrony on windows (not sure how or why) but it should still be able to attempt reading the socket.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (MovieStoreGuy) [reply receiver/chronyreceiver/internal/chrony/client.go:65]: This comes from the original `collectd/chronyd` client where a random sequence value was used.

I am happy to try with time considering the linter would complain less here.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (MovieStoreGuy) [reply receiver/chronyreceiver/internal/chrony/client.go:86]: That is a good point, I did forget to check this. 

In its current state now but if seeding function works with time then it should be absolutely fine.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (jpkrohling) [inline receiver/chronyreceiver/internal/chrony/client.go:89]: There's a slight chance that two concurrent calls at the same nanosecond would execute: what are the implications? Will it work just fine?

**Review by jpkrohling — APPROVED** [review]
just a clarification needed on one part, but LGTM otherwise.

File: receiver/chronyreceiver/internal/chrony/client.go
Comment (MovieStoreGuy) [reply receiver/chronyreceiver/internal/chrony/client.go:89]: Let me double check this by freezing the passed clock with a running `chronyd` instance

**Comment by MovieStoreGuy** [comment]
So I tried running against a local `chronyd` instance on my machine with the following code:

```golang
func main() {
	client, err := chrony.New("udp://127.0.0.1:323", 5*time.Second)
	if err != nil {
		panic(err)
	}
	clk := clock.NewMock(time.Now())
	ctx := clock.Context(context.Background(), clk)

	for i := 0; i < 2; i++ {
		data, err := client.GetTrackingData(ctx)
		if err != nil {
			panic(err)
		}
		fmt.Println("Data ingested:", data)
	}
}
```

There was absolutely no issues returning the data since the clock had not progressed during these two calls.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**
